File: src/pod2spm/detect.py
"""Detect whether a pod ships prebuilt XCFrameworks or needs building from source."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path


@dataclass
class DetectionResult:
    """Result of scanning Pods/ after `pod install`."""

    xcframeworks: list[Path] = field(default_factory=list)
    resource_bundles: list[Path] = field(default_factory=list)

    @property
    def is_prebuilt(self) -> bool:
        return len(self.xcframeworks) > 0


def scan_pods_dir(pods_dir: Path, pod_name: str) -> DetectionResult:
    """Scan the Pods/ directory for xcframeworks and resource bundles.

    Looks for:
      - *.xcframework anywhere under Pods/
      - *.bundle directories (resource bundles with PrivacyInfo.xcprivacy etc.)
    """
    result = DetectionResult()

    if not pods_dir.exists():
        return result

    # Scan for xcframeworks — skip Pods/Target Support Files
    for item in pods_dir.rglob("*.xcframework"):
        if "Target Support Files" not in str(item):
            result.xcframeworks.append(item)

    # Scan for resource bundles
    for item in pods_dir.rglob("*.bundle"):
        if item.is_dir() and "Target Support Files" not in str(item):
            result.resource_bundles.append(item)

    return result

File: lib/pod2spm/detect.rb
# frozen_string_literal: true

module Pod2SPM
  # Result of scanning Pods/ after `pod install`.
  DetectionResult = Struct.new(:xcframeworks, :resource_bundles, :vendored_frameworks) do
    def initialize(xcframeworks: [], resource_bundles: [], vendored_frameworks: [])
      super(xcframeworks, resource_bundles, vendored_frameworks)
    end

    def prebuilt?
      !xcframeworks.empty?
    end
  end

  module Detect
    # Scan the Pods/ directory for xcframeworks, vendored frameworks, and
    # resource bundles scoped to a specific pod.
    #
    # When pod_name is given the search is restricted to Pods/<pod_name>/.
    # Falls back to a full Pods/ scan only when no pod-specific directory
    # exists (e.g. umbrella pods that don't have their own subdirectory).
    #
    # Also detects plain .framework bundles (vendored_frameworks) and
    # returns them separately so callers can warn or handle them.
    def self.scan(pods_dir, pod_name = nil)
      result = DetectionResult.new

      return result unless File.directory?(pods_dir)

      search_root = pods_dir
      if pod_name
        pod_dir = File.join(pods_dir, pod_name)
        search_root = pod_dir if File.directory?(pod_dir)
      end

      xcframeworks = Dir.glob(File.join(search_root, "**", "*.xcframework"))
        .select { |p| File.directory?(p) }
        .reject { |p| p.include?("Target Support Files") }
        .sort

      vendored = Dir.glob(File.join(search_root, "**", "*.framework"))
        .select { |p| File.directory?(p) }
        .reject { |p| p.include?("Target Support Files") }
        .reject { |p| p.include?(".xcframework") }
        .sort

      bundles = Dir.glob(File.join(search_root, "**", "*.bundle"))
        .select { |p| File.directory?(p) }
        .reject { |p| p.include?("Target Support Files") }
        .sort

      result.xcframeworks        = xcframeworks
      result.vendored_frameworks = vendored
      result.resource_bundles    = bundles
      result
    end
  end
end
